File: src/nifs/monitor.zig
const std = @import("std");
const assert = std.debug.assert;
const e = @import("../erl_nif.zig");
const rl = @import("../raylib.zig");

const core = @import("../core.zig");

pub const exported_nifs = [_]e.ErlNifFunc{
    // Monitor
    .{ .name = "set_window_monitor", .arity = 1, .fptr = nif_set_window_monitor, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_count", .arity = 0, .fptr = nif_get_monitor_count, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_current_monitor", .arity = 0, .fptr = nif_get_current_monitor, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_position", .arity = 1, .fptr = nif_get_monitor_position, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_position", .arity = 2, .fptr = nif_get_monitor_position, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_width", .arity = 1, .fptr = nif_get_monitor_width, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_height", .arity = 1, .fptr = nif_get_monitor_height, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_physical_width", .arity = 1, .fptr = nif_get_monitor_physical_width, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_physical_height", .arity = 1, .fptr = nif_get_monitor_physical_height, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_refresh_rate", .arity = 1, .fptr = nif_get_monitor_refresh_rate, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "get_monitor_name", .arity = 1, .fptr = nif_get_monitor_name, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
};

///////////////
//  Monitor  //
///////////////

/// Set monitor for the current window
///
/// raylib.h
/// RLAPI void SetWindowMonitor(int monitor);
fn nif_set_window_monitor(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    rl.SetWindowMonitor(monitor);

    // Return

    return core.Atom.make(env, "ok");
}

/// Get number of connected monitors
///
/// raylib.h
/// RLAPI int GetMonitorCount(void);
fn nif_get_monitor_count(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    const monitor_count = rl.GetMonitorCount();

    // Return

    return core.Int.make(env, monitor_count);
}

/// Get current monitor where window is placed
///
/// raylib.h
/// RLAPI int GetCurrentMonitor(void);
fn nif_get_current_monitor(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    const current_monitor = rl.GetCurrentMonitor();

    // Return

    return core.Int.make(env, current_monitor);
}

/// Get specified monitor position
///
/// raylib.h
/// RLAPI Vector2 GetMonitorPosition(int monitor);
fn nif_get_monitor_position(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1 or argc == 2);

    // Return type

    const return_resource = core.must_return_resource(env, argc, argv, 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const position = rl.GetMonitorPosition(monitor);
    defer if (!return_resource) core.Vector2.free(position);

    // Return

    return core.maybe_make_struct_as_resource(core.Vector2, env, position, return_resource) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Failed to get return value.");
    };
}

/// Get specified monitor width (current video mode used by monitor)
///
/// raylib.h
/// RLAPI int GetMonitorWidth(int monitor);
fn nif_get_monitor_width(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const monitor_width = rl.GetMonitorWidth(monitor);

    // Return

    return core.Int.make(env, monitor_width);
}

/// Get specified monitor height (current video mode used by monitor)
///
/// raylib.h
/// RLAPI int GetMonitorHeight(int monitor);
fn nif_get_monitor_height(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const monitor_height = rl.GetMonitorHeight(monitor);

    // Return

    return core.Int.make(env, monitor_height);
}

/// Get specified monitor physical width in millimetres
///
/// raylib.h
/// RLAPI int GetMonitorPhysicalWidth(int monitor);
fn nif_get_monitor_physical_width(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const monitor_width = rl.GetMonitorPhysicalWidth(monitor);

    // Return

    return core.Int.make(env, monitor_width);
}

/// Get specified monitor physical height in millimetres
///
/// raylib.h
/// RLAPI int GetMonitorPhysicalHeight(int monitor);
fn nif_get_monitor_physical_height(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const monitor_height = rl.GetMonitorPhysicalHeight(monitor);

    // Return

    return core.Int.make(env, monitor_height);
}

/// Get specified monitor refresh rate
///
/// raylib.h
/// RLAPI int GetMonitorRefreshRate(int monitor);
fn nif_get_monitor_refresh_rate(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const monitor_refresh_rate = rl.GetMonitorRefreshRate(monitor);

    // Return

    return core.Int.make(env, monitor_refresh_rate);
}

/// Get the human-readable, UTF-8 encoded name of the specified monitor
///
/// raylib.h
/// RLAPI const char *GetMonitorName(int monitor);
fn nif_get_monitor_name(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const monitor = core.Int.get(env, argv[0]) catch |err| {
        return core.raise_exception(e.allocator, env, err, @errorReturnTrace(), "Invalid argument 'monitor'.");
    };

    // Function

    const monitor_name = rl.GetMonitorName(monitor);
    // Do NOT free monitor_name

    // Return

    return core.CString.make_c_unknown(env, monitor_name);
}

File: lib/zexray/nif/monitor.ex
defmodule Zexray.NIF.Monitor do
  @moduledoc false

  defmacro __using__(_opts) do
    quote do
      @nifs_monitor [
        # Monitor
        set_window_monitor: 1,
        get_monitor_count: 0,
        get_current_monitor: 0,
        get_monitor_position: 1,
        get_monitor_position: 2,
        get_monitor_width: 1,
        get_monitor_height: 1,
        get_monitor_physical_width: 1,
        get_monitor_physical_height: 1,
        get_monitor_refresh_rate: 1,
        get_monitor_name: 1
      ]

      #############
      #  Monitor  #
      #############

      @doc """
      Set monitor for the current window

      ```c
      // raylib.h
      RLAPI void SetWindowMonitor(int monitor);
      ```
      """
      @doc group: :window
      @spec set_window_monitor(monitor :: integer) :: :ok
      def set_window_monitor(_monitor), do: :erlang.nif_error(:undef)

      @doc """
      Get number of connected monitors

      ```c
      // raylib.h
      RLAPI int GetMonitorCount(void);
      ```
      """
      @doc group: :window
      @spec get_monitor_count() :: integer
      def get_monitor_count(), do: :erlang.nif_error(:undef)

      @doc """
      Get current monitor where window is placed

      ```c
      // raylib.h
      RLAPI int GetCurrentMonitor(void);
      ```
      """
      @doc group: :window
      @spec get_current_monitor() :: integer
      def get_current_monitor(), do: :erlang.nif_error(:undef)

      @doc """
      Get specified monitor position

      ```c
      // raylib.h
      RLAPI Vector2 GetMonitorPosition(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_position(
              monitor :: integer,
              return :: :value | :resource
            ) :: tuple
      def get_monitor_position(
            _monitor,
            _return \\ :value
          ),
          do: :erlang.nif_error(:undef)

      @doc """
      Get specified monitor width (current video mode used by monitor)

      ```c
      // raylib.h
      RLAPI int GetMonitorWidth(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_width(monitor :: integer) :: integer
      def get_monitor_width(_monitor), do: :erlang.nif_error(:undef)

      @doc """
      Get specified monitor height (current video mode used by monitor)

      ```c
      // raylib.h
      RLAPI int GetMonitorHeight(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_height(monitor :: integer) :: integer
      def get_monitor_height(_monitor), do: :erlang.nif_error(:undef)

      @doc """
      Get specified monitor physical width in millimetres

      ```c
      // raylib.h
      RLAPI int GetMonitorPhysicalWidth(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_physical_width(monitor :: integer) :: integer
      def get_monitor_physical_width(_monitor), do: :erlang.nif_error(:undef)

      @doc """
      Get specified monitor physical height in millimetres

      ```c
      // raylib.h
      RLAPI int GetMonitorPhysicalHeight(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_physical_height(monitor :: integer) :: integer
      def get_monitor_physical_height(_monitor), do: :erlang.nif_error(:undef)

      @doc """
      Get specified monitor refresh rate

      ```c
      // raylib.h
      RLAPI int GetMonitorRefreshRate(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_refresh_rate(monitor :: integer) :: integer
      def get_monitor_refresh_rate(_monitor), do: :erlang.nif_error(:undef)

      @doc """
      Get the human-readable, UTF-8 encoded name of the specified monitor

      ```c
      // raylib.h
      RLAPI const char *GetMonitorName(int monitor);
      ```
      """
      @doc group: :window
      @spec get_monitor_name(monitor :: integer) :: binary
      def get_monitor_name(_monitor), do: :erlang.nif_error(:undef)
    end
  end
end
